### ISSUE TITLE
Remove some SHM protocols when switching to Nemesis

### ISSUE THREAD

**Issue by a-szegel** [ofiwg/libfabric#8435]
The current SHM provider internal protocols are as follows:
```
enum {
    smr_src_inline, /* command data */
    smr_src_inject, /* inject buffers */
    smr_src_iov,    /* reference iovec via CMA */
    smr_src_mmap,   /* mmap-based fallback protocol */
    smr_src_sar,    /* segmentation fallback protocol */
    smr_src_ipc,    /* device IPC handle protocol */
    smr_src_max,    /* ENUM sentinal, not protocol */
};
```
with a simplified selection logic as follows:
```
if (use_ipc) do_ipc
if (size < 196) do_inline
if (size < 4096) do_inject
if (have_cma) do_iov
else do_sar
// smr_src_mmap should never be selected (unless configuration specified sar_size)
```
With the transition to use the Nemesis algorithm, I do not think all of these protocols make sense anymore.  I would like to see the selection logic change to:
```
if (size < fastbox_max && fastbox_avail && fastbox_open) do_fastbox
if (size < free_queue_size) do_nemesis
if (have_large_msg_optimization) do_large_msg_optimization

do_sar
```

Is there any reason why we need to keep the old protocols alive (IPC and mmap)?

**Comment by shefty** [comment]
@aingerson @wzamazon 

inline seems similar to fastbox.  The necessary change is for shm to have separate tagged queues per peer as a first step, which would eliminate most search time.

shm supports memory transfers to/from GPU memory and host memory -- that's IPC.  CMA is not available on many platforms because of security restrictions.  xpmem will be the same - though even less likely to be available.  There are additional optimizations for using platform accelerations (e.g. DSA - data streaming accelerator) when available.   That's coupled with SAR currently and should be used when available (shows significant benefits for large transfers).  I believe mmap is an option used on AWS.

We need details on *how* the alternative algorithms perform transfers.

**Comment by a-szegel** [comment]
All transfers with Nemesis will have to go through the same process:

Sender:
1. Sender pops top entry off LIFO free queue
     a.  (Memory Util Optimization) Multiple free queue's with different message payload sizes
     b.  (Memory Util Optimization) Can dynamically grow the number of elements in a free queue
2. Fills entry with message
3. Write entry into peers recv (cmd) queue
4. Sender can write Transmit Complete event if msg fits in entry payload

Receiver:
1. Poll recv (cmd) queue (and all active recv fast boxes)
2. Copy the contents of the first message somewhere (tagged message, or local bounce buffer)
3. Return used queue to Senders recv queue
    a. When the sender processes the returned queue, it can now write a delivery complete event
    b. We have a natural response mechanism so we can remove the old one

The inline protocol writes directly into the peers circular buffer.  With the Nemesis algorithm, that will no longer be possible b/c the peer will have a fifo queue that will be empty when there are no messages to read. If we choose to implement fast boxes (improved latency by skipping queue), there will be a limited number of them, so we may fall back to using basic nemesis protocol under heavy load.

The inject protocol maps to the Nemesis implementation.  We can have multiple free queue list with payload sizes of 1k, 4k, and 8k so we can potentially extend it further (OMPI goes up to 32k).

SAR will sit on top of the nemesis implementation, segment the message up into whatever size free-queues are available and ship it across.  When you run out of free queues, you will need to call into the progress engine to replenish the free queue.

For read based operations, we will use nemesis to convey address and size information to the peer, and have the peer kick off the copy (this should work for GPU's too, with a ofi_hmem_copy as fallback).

Given the above, I don't really see the need to keep mmap or IPC protocols.

**Comment by shefty** [comment]
IPC is used for transfers involving GPU memory.  CMA avoids extra memory copies to save memory bandwidth.

**Comment by aingerson** [comment]
We need to keep IPC and CMA - it is much more performant to do one copy for larger messages (whether host or GPU memory). Going through the SAR protocol will greatly increase the number of copies.
Mmap I don't think it used by anyone. It was implemented by Amazon as an initial solution to the large message copies without CMA and was replaced by SAR which performs much better.
Moving the SAR on top of nemesis breaks the DSA implementation which uses asynchronous data copies

**Comment by a-szegel** [comment]
Can you give more detail on why DSA will break?

**Comment by aingerson** [comment]
It seems like what you proposed requires the data to be copied into the payload synchronously. DSA uses a separate pool of bounce buffers to copy data asynchronously into there and notify the receiver whenever the next batch of data is ready to be copied out.
I probably am just misunderstanding the assumptions of your proposal though. Do you see it working with asynchronously copied data?

**Comment by shefty** [comment]
To clarify, DSA has hardware-based copy engines, and does not use the CPU to perform the copy.

**Comment by a-szegel** [comment]
I think the DSA case can work similarly to other READ based cases.  We use Nemesis to tell the receiving process that it can call the DSA to transfer memory (Read memory from sending process to itself), and return the free queue once the transfer is complete.  If the sender needs to kick off the asynch DSA copy and needs to know the destination on the host before initiating the copy, we can add a higher level protocol to take care of that synchronization (while still using the nemesis algorithm for the inter-process synchronization... i.e.. "are you ready to receive", "Yes, please send to addr XXX, size Y", "Done Sending", "Message Received!").

**Comment by aingerson** [comment]
DSA uses the hardware copy engines to parallelize multiple of the segments all at once. For example, if it has 8 segments, it might distribute them across 8 different engines. Or it might put 2 segments each on 4 engines and those 2 will be done in parallel as well. Do you see that design working with your proposal?
The DSA sender does not need to know the destination of the message before initiating the transfer. It is just a bounce buffer protocol. There is a CMA RMA optimization that does this though and it could make use of that but I'm not sure how widely it is used if at all.

**Comment by a-szegel** [comment]
I think the mis-understanding is that we don't always have to put the data in the payload, but I always want to use the nemesis queue to pass the information on "heres how we are going to do the transfer" to the other side (fully rip out all circular queues).

**Comment by aingerson** [comment]
Yes, that was definitely my misunderstanding :)
I thought you were suggesting replacing the SAR implementation with always putting the data in the payload when posting the message and having everything go through this new SAR protocol.
If all you're changing is how those commands are sent and received then I think that is fine. I don't think I would make many adjustments to the protocols though since we have a clear use case for each one (ie DSA, IPC, CMA). The ones I could see use getting rid of are mmap (since it is not as performant and I don't think anyone uses it) and the inject protocol (you could remove the inject pool and just increase the size of the inline cmd payload, increasing the inline threshold).

**Comment by a-szegel** [comment]
I wanted to generalize DSA, IPC, and CMA into one "read based" protocol where everything is the same from the sender side (sender provides addr, size, and op-code), and then the receiver uses that op-code in a switch statement to pick the transfer method.  I need to spend more time in the code to understand exactly what IPC and DSA are doing.

**Comment by aingerson** [comment]
The DSA one is not a separate protocol right now. It just goes through the SAR protocol and bounces through the SAR buffers, using the DSA engines. Both sides also don't need to use DSA ie the receiver doesn't care what the sender used, it will copy the data from the SAR buffers however it wants.
The IPC and CMA protocols I think will be tricky to merge into one and that's because the sender decides which protocol to take and has to format the addressing information according to that protocol, letting the sender know how to interpret the data. For the IPC protocol, the sender has to create an IPC handle (specific to the interface that ones the memory - cuda, ze, rocr, etc) and lets the receiver know that it is an IPC handle and what interface owns that handle. The receiver then opens that handle and uses the appropriate interface to copy that memory from the sender.
The CMA protocol also uses a read but the information it sends to the sender is inherently different and the receiver needs to know that. It sends the virtual address and the receiver just does a CMA read on that virtual address.
There is a chance to merge them by specifying that CMA is the FI_HMEM_SYSTEM method for IPC were the handle is just the virtual address (and get/open/close handle are a no-op). The challenge there is with the incoming support for XPMEM https://github.com/ofiwg/libfabric/pull/8200. In this case, FI_HMEM_SYSTEM has two different methods for IPC copies and we will have to figure out how to differentiate between those two on the sender side.

**Comment by a-szegel** [comment]
Thank you!  You have convinced me to be minimally invasive to SAR, IPC, and IOV.
